source/geod24/concurrency.d

```diff
@@ -1563,3 +1564,168 @@ private:
     self.receiveOnly!(bool);
     assert(x[0] == 5);
 }
+
+
+/*******************************************************************************
+
+    This is similar to go buffered channels
+
+*******************************************************************************/
+
+public class Channel (T)
+{
+    /// closed
+    private bool closed;
+
+    /// lock for queue and status
+    private Mutex mutex;
+
+    /// queue of data
+    private DList!T queue;
+
+    /// Ctor
+    public this ()
+    {
+        this.closed = false;
+        this.mutex = new Mutex;
+    }
+
+
+    /***************************************************************************
+
+        Put data `msg`.
+
+        Params:
+            msg = message to send
+
+        Return:
+            If successful, return true, otherwise return false
+
+    ***************************************************************************/
+
+    public bool send (T msg)
+    {
+        this.mutex.lock();
+        scope (exit)
+            this.mutex.unlock();
+
+        if (this.closed)
+            return false;
+
+        this.queue.insertBack(msg);
+        return true;
+    }
+
+
+    /***************************************************************************
+
+        Return the message.
+
+        Return:
+            msg = message to receive
+
+    ***************************************************************************/
+
+    public T receive ()
+    {
+        T msg;
+        while (!tryReceive(&msg)) {
+            FiberScheduler.yield();
+        }
+        return msg;
+    }
+
+
+    /***************************************************************************
+
+        Return the message.
+
+        Params:
+            msg = point of message to send
+
+        Return:
+            If successful, return true, otherwise return false
+
+    ***************************************************************************/
+
+    public bool tryReceive (T *msg)
+    {
+        this.mutex.lock();
+        scope (exit)
+            this.mutex.unlock();
+
+        if (this.closed)
+            return false;
+
+        if (this.queue[].walkLength > 0)
+        {
+            *(msg) = this.queue.front;
+            this.queue.removeFront();
+            return true;
+        }
+
+        return false;
+    }
+
+
+    /***************************************************************************
+
+        Return closing status
+
+        Return:
+            If channel is closed, return true, otherwise return false
+
+    ***************************************************************************/
+
+    public @property bool isClosed () @safe @nogc pure
```

Comment:
Why is this the only method that is `@safe` / `@trusted`? That means other `@safe` code can only use this function in this class.